reps/2023-8-18-ray-on-spark-autoscaling.md

```diff
@@ -0,0 +1,270 @@
+
+
+#### Shall we merge multiple Ray worker nodes into one spark job ?
+
+This is difficult. When we cancel a spark job, all spark tasks in the job are killed.
+But the `NodeProvider` interface might trigger `terminate_node` solely on specific
+Ray node.
+
+On the other hand, create multiple spark jobs, one job with only one spark task
+(each spark task holds a Ray worker node)
+does not cause too much overhead, assuming we don't have enormous number (< 10000)
+of Ray worker nodes.
+
+
+#### Can we create a new spark session in Ray autoscaler process instead of reusing parent process spark session ? So that we can avoid create a spark job server.
+
+This way it means creating a separate spark application but not running Ray cluster
+in parent process spark application, but the `setup_ray_cluster` is designed to
+start a Ray cluster via current spark application, and all Ray cluster resources
+requests must be limited inside the current spark application.
+
+So that we cannot use this way.
+
+#### When `NodeProvider.create_node_with_resources_and_labels` is triggered, and node creation request is emitted but it takes several minutes to launch a new spark worker nodes and create ray worker node, during the creation pending period, how to notify Ray autoscaler that the Ray worker node is pending creation, but not creation failure ?
+
+To be answered.
+
+
+#### When Ray worker nodes crashes unexpectedly, autoscaler should handle the case properly and start replacement nodes if needed. But we use spark task to hold the worker node, if the spark task fails, spark will retry running failed task. How to address conflicts here ?
+
+
+In spark task, we start a child process to execute Ray worker node and the spark task python worker wait the child process until it terminates. When child processs (ray worker node) exit unexpectedly, the parent process (spark task python worker) does not raise error but returns normally, then spark will regard the task completes normally and does not trigger task retry but complete the spark job.
+
+
+#### Ray autoscaler supports setting multiple Ray worker groups, each Ray worker group has its individual CPU / GPU / memory resources configuration, and its own minumum / maximum worker number setting for autoscaling. Shall we support this feature ?
+in Ray on spark autoscaling ?
+
+We can support this if customer requires it.
+
+
+#### What's the default Ray on spark minimum worker number we should use ?
+
+I propose to set it to zero. Ray worker node launching is very quick, and setting it to
+zero increases cluster resources utilization ratio especially when the Ray on spark cluster is often idle.
+
+
+#### Apache/spark dynamic resource allocation also supports Spark over Kubernetes/Yarn, can we make Ray on spark autoscaling supports Spark over Kubernetes/Yarn ?
+
```

Comment:
To add a section to explain:

why don't use `ray up` to start ray cluster with autoscaling ?